### ISSUE TITLE
tf.py_func returned object doesn't supported in tf.train.shuffle_batch

### ISSUE THREAD

**Issue by dmitryfisko** [tensorflow/tensorflow#2193]
```
def unpackbits(arr):
    return np.unpackbits(arr).astype(np.int64)

labels, = tf.py_func(unpackbits, [labels], [tf.int64])
```

I use py_func for reading input data. Returned object from tf.py_func(...) has get_shape()._dim equal to None. Its raise exception when passing to tf.train.shuffle_batch

**Comment by yaroslavvb** [comment]
Use `labels.set_shape(())`

`batch` is a kind of FIFOQueue and it needs to know sizes of inputs for memory allocation reasons. It can either get it from `shapes` argument, or use shape inference to infer it automatically. Since `py_func` can call arbitrary Python code and give different-shaped argument between invocations, `py_func` doesn't set static shape and you need to either use `shapes` argument to `batch` or `set_shape` on `labels` tensor.

So you either need to 

`labels.set_shape(())`

or do

`labels_batch = tf.batch(..., shapes=[()])`

Here's an end-to-end example to test this

```
# Create queue, initialize it with 0,1,2,3,4,5,6,7,8,9
queue_size = 10
tf.reset_default_graph()
queue = tf.FIFOQueue(capacity=queue_size, dtypes=tf.int64)
enqueue_placeholder = tf.placeholder(dtype=tf.int64)
enqueue_op = queue.enqueue(enqueue_placeholder)
dequeue_op = queue.dequeue()

sess = tf.InteractiveSession()
for f in range(queue_size):
  sess.run([enqueue_op], feed_dict={enqueue_placeholder: f})
sess.run(queue.close())

def unpackbits(arr):
    return arr.astype(np.int64)

labels = dequeue_op
labels2, = tf.py_func(unpackbits, [labels], [tf.int64])

labels2.set_shape(())
labels2_batch = tf.batch([labels2], 2, capacity=10)

tf.start_queue_runners()

for i in range(5):
  print labels2_batch.eval()
```

**Comment by dmitryfisko** [comment]
Thank you.
